underscore: Expand types for zip and friends

LukeGT opened the pull request:
This adds more specific type overloads to zip, unzip and transpose methods. It caters to the typical usecase where 3 or less lists are passed in, or when the contents of all passed lists have the same type.

The output types are technically wrong in situations where the input lists are of different lengths. In these cases the returned list might contain undefined elements.

I could add optionality to the types where appropriate, but this would not be in line with Typescript's own opinions on a similar concept with respect to accessing arrays - Typescript assumes that array access returns an object of the array's type without adding `undefined`, even though `undefined` is a possible value. There is a flag to account for these undefined accesses, but it's not even part of the "strict' Typescript setting. https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-1.html#checked-indexed-accesses---nouncheckedindexedaccess

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://underscorejs.org/#zip, https://underscorejs.org/#unzip
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

LukeGT commented:
I'm looking for advice on how to resolve an issue where Typescript 4.6 and below arbitrarily swaps the order of a union from `string[] | number[]` to `number[] | string[]`, causing the tests to fail. The types are equivalent, but the order makes the test framework think they are different.

I can't raise the minimum Typescript requirement to 4.6 because `backbone` requires a lower Typescript version, and I feel like changing that is going to be a huge headache (but feel free to correct me on that).

Is there a way to have `$ExpectType` disregard union order, or to have different assertions for different versions of Typescript?

My current workaround is just to comment out those test lines, which is far from ideal, but basically kicks the can down the road until February 2024 when Typescript 4.6 exits the support window.

Very much open to any other suggestions you might have :)

jgonggrijp commented on types/underscore/index.d.ts:
```diff
@@ -859,6 +860,11 @@ declare module _ {
          * @returns The zipped version of `lists`.
          **/
+        zip(): [];
+        zip<T, U, V>(...lists: [List<T>, List<U>, List<V>]): [T, U, V][];
+        zip<T, U>(...lists: [List<T>, List<U>]): [T, U][];
```
Depending on how well TS is able to infer the types of lists, these two declarations might seriously break someone's code (because of the unequal length problem, as you mentioned yourself). Is there an easy way in which users can opt back out to the `List<any>[]` case?

LukeGT replied:
They could invoke `zip<any>(...)` instead if they wanted to opt out of stricter types.

I wonder though what code would be broken by this change? I can only think of situations where upgrading to this type definition would reveal a real bug. I guess this is a question around what you mean by "broken"?

For example, in the situation where lists of unequal length are provided, the return type should really be `[T?, U?][]`, which would force the user to check for undefined values before attempting to use the values for `T` and `U`. The risk of returning `[T, U][]` instead is that the user is not prompted to check for `undefined`, though they are welcome to do so (it is not a type error). But at least TS knows that the two zipped values should be of type `T` and `U` respectively.

In the current implementation, the return type is `any[][]`, so no type safety is given at all, let alone a reminder to check for `undefined` values.

jgonggrijp replied:
I tend to look at these things from the JS perspective. If I know that the arrays may be of unequal length, then I am aware that tuples at the end of the output array might contain `undefined` and I am probably writing code that takes this into account. With a TS typing of `any[][]`, this makes sense. If the TS typing changes to `[T, U][]` (where `T` and `U` exclude `undefined`), I suddenly find myself in disagreement with the TSC (and the TSC is technically wrong).

This might go by unnoticed (I guess this depends on strictness settings), but it could very well result in a warning or even an error down the line, which means that the TSC is going to tell me that my code is wrong when it is not.

That said, `zip<any>(...)` fits the bill of an opt out perfectly.

LukeGT replied:
That makes sense, thanks for elaborating!

How worried are you about future TS changes causing breakages vs present ones?

If it's just about the present, then I can't seem to find an example where working JS code results in TS compilation errors, even with strict turned on. My best shot is below. Within the `if` block, `a` adopts the `never` type, but TS doesn't throw an error even if you try to use the resulting value in an `undefined` context.

```
let c: undefined = undefined;
for (const [a, b] of _.zip([1, 2, 3], [4, 5])) {
    if (a === undefined) {
        c = a;
    }
}
```

If you're worried about the future, then I think TS will most likely not break this behaviour because of how they treat arrays. The following code also results in a `never` type, for example:

```
const array = [1, 2, 3];
const el = array[4];
if (el === undefined) {
  // el is never
}
```

I can also write a test to make sure that using zip with variable length lists does not break.

If I add such a test, combined with the reasoning above, do you think that's enough to address your concerns?

jgonggrijp replied:
I haven't looked at the latest changes yet, so what I write here might be outdated. With that out of the way:

> How worried are you about future TS changes causing breakages vs present ones?

Equally. 😉 

> 
> If it's just about the present, then I can't seem to find an example where working JS code results in TS compilation errors, even with strict turned on.

I can't think of anything off the top of my head, but I know that in my daily work, I run into these situations all the time. It's usually something subtle, indirect and unexpected and part of the problem is usually in DT packages. I don't blame the authors; JavaScript is dynamically typed, so it is impossible to do a perfect job with static type annotations, and the work is usually done for free. Still, the fact that part of the problem is in a DT package tends to make workarounds really fiddly and cumbersome. That, or I have to resort to `any`.

So, while I cannot always foresee how a subtly incorrect typing like this one will cause problems, I very much expect it to.

I take your word on the examples you wrote, but to be honest, I find them a bit simplistic; they don't seem to be representative of the problems I usually encounter.

> I can also write a test to make sure that using zip with variable length lists does not break.
> 
> If I add such a test, combined with the reasoning above, do you think that's enough to address your concerns?

Yes and no. I'm honestly a bit pessimistic that the changes won't cause problems, but with the easy escape that `zip<any>` provides, I accept the possibility. I have to write workarounds after minor and patch updates in DT packages all the time, so I consider it a fact of life.

LukeGT replied:
Thanks for the detailed reply, that makes a lot of sense, I'm sure it's a wild west out there. Fingers crossed this doesn't cause too much trouble, but as you say, there's an escape hatch.

jgonggrijp commented on types/underscore/index.d.ts:
```diff
@@ -869,8 +875,18 @@ declare module _ {
          **/
+        unzip<T, U, V>(lists: List<[T, U, V]>): [T[], U[], V[]];
+        unzip<T, U>(lists: List<[T, U]>): [T[], U[]];
+        unzip<T>(lists: List<[T]>): [T[]]; // eslint-disable-line no-single-element-tuple-type
+        unzip<T>(lists: List<List<T>>): T[][];
```
Don't these two lines mean roughly the same? Why would you need both?

LukeGT replied:
The first line means "a list of length 1 that contains a list of T", whereas the second line means "a list of lists of T". The first type would catch an error where a user tries to do a destructuring assignment into too many variables, for example:

```
const [a, b] = _.unzip([[1], [2]]); // Tuple type '[number]' of length '1' has no element at index '1'. (2493)
const [a, b] = _.unzip([[1, true], [2, false]]); // all is well
```

jgonggrijp commented on types/underscore/index.d.ts:
```diff
@@ -4621,8 +4641,24 @@ declare module _ {
          * the second elements, and so on. (alias: transpose)
          * @returns The unzipped version of the wrapped lists.
          **/
-        unzip(): any[][];
-        transpose(): any[][];
+        unzip(): T extends [infer A, infer B, infer C]
+            ? [A[], B[], C[]]
+            : T extends [infer A, infer B]
+            ? [A[], B[]]
+            : T extends [infer A] // eslint-disable-line no-single-element-tuple-type
+            ? [A[]] // eslint-disable-line no-single-element-tuple-type
+            : T extends List<infer A>
+            ? A[][]
+            : never[];
+        transpose(): T extends [infer A, infer B, infer C]
+            ? [A[], B[], C[]]
+            : T extends [infer A, infer B]
+            ? [A[], B[]]
+            : T extends [infer A] // eslint-disable-line no-single-element-tuple-type
+            ? [A[]] // eslint-disable-line no-single-element-tuple-type
+            : T extends List<infer A>
+            ? A[][]
+            : never[];
```
If I understand correctly, this is a destructive change; unlike the other changes, there is no fallback to `any[][]`.

jgonggrijp commented on types/underscore/index.d.ts:
```diff
@@ -5862,8 +5902,24 @@ declare module _ {
          * @returns A chain wrapper around the unzipped version of the wrapped
          * lists.
          **/
-        unzip(): _Chain<any[]>;
-        transpose(): _Chain<any[]>;
+        unzip(): T extends [infer A, infer B, infer C]
+            ? _Chain<A[] | B[] | C[], [A[], B[], C[]]>
+            : T extends [infer A, infer B]
+            ? _Chain<A[] | B[], [A[], B[]]>
+            : T extends [infer A] // eslint-disable-line no-single-element-tuple-type
+            ? _Chain<A[], [A[]]> // eslint-disable-line no-single-element-tuple-type
+            : T extends List<infer A>
+            ? _Chain<A[]>
+            : _Chain<never>;
+        transpose(): T extends [infer A, infer B, infer C]
+            ? _Chain<A[] | B[] | C[], [A[], B[], C[]]>
+            : T extends [infer A, infer B]
+            ? _Chain<A[] | B[], [A[], B[]]>
+            : T extends [infer A] // eslint-disable-line no-single-element-tuple-type
+            ? _Chain<A[], [A[]]> // eslint-disable-line no-single-element-tuple-type
+            : T extends List<infer A>
+            ? _Chain<A[]>
+            : _Chain<never>;
```
Same seemingly-destructive change here.

LukeGT replied:
Ditto reasoning above

jgonggrijp reviewed (commented):
Maybe you can define a helper type somewhere as `string[] | number[]` and use that everywhere instead of the literal union? In that way, you might be able to abstract out the arbitrary swap.

I'm rusty on TS, so please forgive me for any naieve questions. Once I get a better grasp of the changes, I'll update the review verdict to "approve" or "request changes".

LukeGT commented:
I'll have a go at aliasing the union, that could work if I pass it in as a type argument manually! Just have to pray that TS is not too clever for its own good.

I would also like some clarity on what you consider to be "broken" when making changes here. Yes, these changes might break someone's build if they upgrade, but only if there are actual bugs in their code (assuming I've done my due diligence in testing).

jgonggrijp commented:
If my code is correct *from the JS perspective*, and TSC accepts my code before the change but not after it, then I consider it a breaking change.

I know this is an impopular opinion among TS adepts, but I stand by the principle that TS should help me discover real errors, not make my job harder by having an opinion about which parts of JS are "cool" and which are "uncool".

LukeGT commented:
I definitely agree with you in principle, I just hope to convince you that the types I've written will strictly help people find errors rather than create false positives.

I've made some changes, please take a look.

- Fixed the two tests that were broken on Typescript 4.6 previously, your suggestion around using an alias worked
- Corrected `never[]` to `[]` (explicitly the empty list)
- Added a test to make sure that testing for undefined doesn't break compilation, to cover the variable length list case

jgonggrijp commented on types/underscore/underscore-tests.ts:
```diff
@@ -2459,6 +2459,17 @@ _.zip(['moe', 'larry', 'curly'], [30, 40, 50], [true, false, false]); // $Expect
+        if (right === undefined) {
+            const right_check: undefined = right;
+        }
+    }
```
Yes, this test is somewhat reassuring, if not entirely. Thanks for adding it.

jgonggrijp commented on types/underscore/underscore-tests.ts:
```diff
@@ -2500,9 +2513,11 @@ _.zip(['moe', 'larry', 'curly'], [30, 40, 50], [true, false, false]); // $Expect
-    // extractChainTypes(_.chain(tupleList).transpose()); // $ExpectType ChainType<[string[], number[]], string[] | number[]>
+    // Typescript 4.6 and below arbitrarily swaps the string[] | number[] union to number[] | string[],
+    // so we create a type alias to normalise the union.
+    type StringOrNumberList = string[] | number[];
```
Why not lift the definition from line 2483 so you can reuse it here?

LukeGT replied:
Fair point, done.

jgonggrijp reviewed (approved):
> I definitely agree with you in principle, I just hope to convince you that the types I've written will strictly help people find errors rather than create false positives.

Oh, just for clarity, I do think that your types are an improvement. It helps if TS understands that the first element of a tuple is likely to be a string and the second element isn't.

I'm not entirely convinced yet that the changes don't come with some new false positives, but as I expressed before, I accept that as long as there is an easy workaround.

>     * Fixed the two tests that were broken on Typescript 4.6 previously, your suggestion around using an alias worked

Glad I could help there.

>     * Corrected `never[]` to `[]` (explicitly the empty list)

That does look less scary, even if it doesn't matter as much as I thought initially. Thanks.

I recommend deduplicating `StringOrNumberList`, but it is not a condition for acceptance.

LukeGT commented:
Fair enough, it's certainly difficult to foresee all possible valid uses in the wild.

I thought of one more edge case, where through `_(...)` or `_.chain(...)` a user is able to pass a non-list, which underscore always returns `[]` for. Let me know how this final change looks to you.

jgonggrijp reviewed (approved):
This looks very good to me! I think you just made the types better at catching errors and I don't smell any new opportunities for false positives. 👍

LukeGT commented:
Great, thanks so much [jgonggrijp](https://github.com/jgonggrijp) for the review!

LukeGT commented:
Ready to merge

mness-cc commented:
Late to the party, but thought I'd just toss in a mention that this also looks good to me, thanks @LukeGT!

I also wanted to note that whenever it's feasible to raise the minimum TS version to TS 3.1, I think we can potentially take advantage of the [mapped Tuple types](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-1.html#mapped-types-on-tuples-and-arrays) feature to collapse these overloads and support theoretically infinte-length tuple types (TS probably has some practical limit though).

![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/121688494/e40f7dad-68fd-47eb-a6e0-ff498968d46c)
![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/121688494/4e34df41-2401-4f54-9abc-223e775447b9)
![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/121688494/4f0b7aaf-7df5-49e2-9fc7-31dde178a74d)

One caveat here for `unzip` at least is that things get hairy if the tuples aren't told explicitly that they're tuples, which could be considered a potential breaking change and might be a good reason to stick with the current finite set of overloads; I'm not sure that theres' a good workaround for that.
![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/121688494/29c4707c-e055-4d66-b0e1-aca1121675d3)

reubenrybnik commented:
Sorry, I forgot I have multiple accounts these days and was logged in with the wrong account for the above comment, this is the account that I should have used.
